tests/test_2qbits_gates.py

```diff
@@ -0,0 +1,162 @@
+from perceval import BasicState, generate_all_logical_states, SimulatorFactory, Processor
+from perceval.components import catalog, BS, get_basic_state_from_ports
+from perceval.algorithm import Analyzer
+
+
+def test_fidelity_and_performance_cnot():
+    # Tests the performance and the fidelity of different CNOT in perceval
+    ports = [pcvl.Port(pcvl.Encoding.DUAL_RAIL, "")] * 2
+    state_dict = {pcvl.components.get_basic_state_from_ports(ports, state): str(
+        state) for state in pcvl.utils.generate_all_logical_states(2)}
+    expected = {"00": "00", "01": "01", "10": "11", "11": "10"}
+
+    # KLM CNOT
+    analyzer_klm_cnot = Analyzer(catalog["klm cnot"].build_processor(), state_dict)
+    analyzer_klm_cnot.compute(expected=expected)
+
+    assert pytest.approx(analyzer_klm_cnot.fidelity, 1E-4) == 1
+
+    # Postprocessed CNOT
+    analyzer_postprocessed_cnot = Analyzer(catalog["postprocessed cnot"].build_processor(), state_dict)
+    analyzer_postprocessed_cnot.compute(expected=expected)
+
+    assert analyzer_postprocessed_cnot.fidelity == 1
+
+    # CNOT using CZ : called - Heralded CNOT
+    analyzer_heralded_cnot = Analyzer(catalog["heralded cnot"].build_processor(), state_dict)
+    analyzer_heralded_cnot.compute(expected=expected)
+
+    assert pytest.approx(analyzer_heralded_cnot.fidelity) == 1
+
+    assert analyzer_postprocessed_cnot.performance > analyzer_heralded_cnot.performance > analyzer_klm_cnot.performance
+
+
+def check_cz_with_heralds(processor, herald_states, error=1E-6):
+    """Check if the cz is correct
+
+    :param processor: CZ processor to check, we assume that ctrl is on [0,1]
+    and data on [2,3]
+    :param herald_states: Basic state corresponding to the heralded or ancillary modes
+    at the end of the circuit
+    :param error: Tolerance for the real part of the prob_amplitude (when not expecting 0), defaults to 1E-6
+    This param represent whether this cz gate is balanced
+    """
+    ports = [pcvl.Port(pcvl.Encoding.DUAL_RAIL, "")] * 2
+    states = [get_basic_state_from_ports(ports, state) * herald_states for state in generate_all_logical_states(2)]
+    sim = SimulatorFactory().build(processor)
+    data_state = BasicState("|0,1,0,1>") * herald_states
+    value = None
+    for i_state in states:
+        for o_state in states:
+            pa = sim.prob_amplitude(i_state, o_state)
+            rpa = np.real(pa)
+            ipa = np.imag(pa)
+            if i_state == o_state:
+                if value is None:
+                    value = np.real(pa)
+                if i_state != data_state:
+                    assert pytest.approx(rpa, error) == value
+                    assert rpa > 0
+                else:
+                    assert pytest.approx(rpa, error) == -value
+                    assert rpa < 0
+                assert pytest.approx(ipa) == 0
+            else:
+                assert pytest.approx(rpa) == 0
+                assert pytest.approx(ipa) == 0
+
```

Comment:
I think that on this whole fonction you are ensuring a special case where the amplitudes are real. Thus you exclude benign global phases that you can have in some optical implementation.  So this works because your amplitudes will be exactly +1/3 or -1/3. 
Be careful: if you have some amplitudes that are +/- 1/3 + ipa and ipa=0, then your gate can pass the test but not be a CZ. 
If you want to be more general ( but still ensuring the CZ works), you should look at the exponential form : r.exp(i.θ) . 
If you have this you have to ensure that:
_ if input==output then r=value else r=0 
_if input==ouput and input != data_state then θ=phase_value 
_if input==ouput and input == data_state then θ=phase_value +pi ( mod 2pi ?)

Reply:
@GdGliniasty, how would a global phase affect? Would it not be factored out?

Reply:
Indeed it would be factored out, but this test doesn't allow it. Let say your gate is diag(i,i,i,-i). This is basically iCZ . For in_state=out_state=|00>, rpa=0 and i=1 then value =0 and and the lign "assert" rpa>0" breaks the test.